backend/Virkarekisteri/src/Utils/VacancyNumberGenerator.cs

```diff
@@ -0,0 +1,44 @@
+using Microsoft.EntityFrameworkCore;
+using Virkarekisteri.Models;
+
+namespace Virkarekisteri.Utils;
+
+public class VacancyNumberGenerator
+{
+    private readonly VirkarekisteriDb db;
+
+    public VacancyNumberGenerator(VirkarekisteriDb db)
+    {
+        this.db = db;
+    }
+
+    /// <summary>
+    /// Generates a unique vacancy number based on the OrgTreeId
+    /// </summary>
+    /// <param name="orgTreeId">The OrgTreeId used to generate the prefix for the vacancy number </param>
+    /// <returns>A unique vacancy number string </returns>
+    public async Task<string> GenerateVacancyNumber(Guid orgTreeId)
+    {
+        // Get the organization number (cost center number / kustannuspaikka) associated with the OrgTreeId
+        var orgNumber = await db
+            .OrganizationTrees.Where(o => o.Id == orgTreeId)
+            .Select(o => o.Number)
+            .FirstOrDefaultAsync();
+
+        string vacancyPrefix = orgNumber ?? "";
+
+        // Get the most recent vacancy number with the same prefix if there is
+        var latestVacancyNumber = await db
+            .Positions.Where(p => p.VacancyNumber.StartsWith(vacancyPrefix))
```

Comment:
Compiler warns that we're referencing a possible null value here (p.VacancyNumber can be null)

You can check for null before accessing the value:  
`.Positions.Where(p => p.VacancyNumber != null && p.VacancyNumber.StartsWith(vacancyPrefix))`